### ISSUE TITLE
Free device before shairport-sync playback

### ISSUE THREAD

**Issue by bar1792** [mikebrady/shairport-sync#464]
Quick explanation, I am using a sound card to act as an aux input for devices that cannot connect otherwise. The way I get audio to play via the aux cable is as follows:
arecord -D plughw:1 -f dat | aplay -D plughw:1 -f dat& 

I have used this with the run_after_playback_ends script and it works fine. However, using
pgrep are | xargs sudo kill 
In the script does not work to kill the aux playback. 
However, it does properly work when run from a terminal.

When shairport runs the script the aux continues to play

**Comment by bar1792** [comment]
Is there another way to free the device in a similar manner or a completely different way?

**Comment by mikebrady** [comment]
It sounds like an issue caused by the difference in the runtime environment between the terminal and running as a script from Shairport Sync. Remember all paths have to be fully specified in a script and permissions will be different. If it runs from a terminal, then it will be possible to make it run from a script.

**Comment by bar1792** [comment]
It's rather odd, but the rest of the script runs just not the first line that kills the PID. I have the rest of the script that sets alsa volume to the proper level. How are these scripts run by shairport? Is it run by another user ?

**Comment by bar1792** [comment]
the script is as follows:
!/bin/bash
/bin/ps -ef | /bin/grep are | /bin/grep -v /bin/grep | /usr/bin/awk '{print $2}'| /usr/bin/xargs /usr/bin/sudo /bin/kill
sleep 5s
if [ $(date +%H) -ge "18" -o $(date +%H) -le "7" ]; then
        amixer set Speaker 100%
else
        amixer -c 1 set Speaker 100%
fi
bash -x /home/pi/shScripts/shairportfade.sh&

exit 0


then the script that is being run after shairport ends is as follows:
#!/bin/sh
amixer set Speaker 70%
arecord -D plughw:1 -f dat | aplay -D plughw:1 -f dat&
exit 0


I have tried changing the shebang from #!/bin/sh to #!/bin/bash but neither of them work

**Comment by mikebrady** [comment]
I just edited the post to make the code look right. I presume there should be a # at the start of the first line.

I'm afraid I don't know what the problem with the `kill` command is – tracing execution of the script might help. Could the `sudo` command be looking for a password? Maybe other readers might have an idea.

**Comment by bar1792** [comment]
thank you, I am still learning github's editing and forgot about that feature. I have edited the script to show the proper shebang notation. Your presumption is correct as the script does include a # on the first line, I simply didn't select it when copying the code. I will try some tracing to see if I can come to some conclusion. Thanks for your quick response

**Comment by joerg-krause** [comment]
You have to send a `SIGINT` (maybe `SIGKILL` works, too) to actually kill `arecord`.

**Comment by bar1792** [comment]
I have tried using sigint as well as sigkill. Unfortunately, neither of them work. The kill command works fine within the terminal, but does not work within the script. I also added a cat command to store user name into a file when executed and that doesn't happen either, but the amixer set works as expected.

**Comment by bar1792** [comment]
Ok I have solved the issue. When shairport-sync is enabled with systemctl and enabled as suggested in the readme.md a script is unable to kill the process for whatever reason. What I was able to do was run:

`sudo systemctl disable shairport-sync`

and add

`/home/pi/shairport-sync/shairport-sync&`

to my rc.local file and now my scripts above work correctly and allow for use of an auxillary cable.


Thanks for all the help, if anyone can explain to me as why this happens I would be extremely appreciative.

**Comment by mikebrady** [comment]
Well done, and I'm afraid I don't know what is going on there...

**Comment by mikebrady** [comment]
@rkam I think the kill target is `arecord`.

**Comment by bar1792** [comment]
@rkam I was not trying to isolate shairport-sync, but rather isolate the PID for arecord or aplay (either will work). As with my second script or the script that runs upon shairport-sync playback ends, it runs arecord and pipes it to aplay for an auxillary input. I didn't want shairport to stop running, but when it is run again I wanted shairport to free up the sound card by killing aplay or arecord. There are multiple ways to kill the process, the simplest being `/bin/pgrep arecord | /usr/bin/xargs /usr/bin/sudo /bin/kill` ,but since that didn't work I when to using several other methods thinking it may have been something to do with pgrep. However, that wasn't the case. The issue somehow stems from shairport-sync being started with systemd upon boot. 

As I was debugging, I killed shairport and ran it from the command line, then when I connected to the device from my phone, the scripts ran properly and killed aplay and arecord. This led me to believe it was an issue with shairport-sync starting with the system daemon. I then issued the command `sudo systemctl disable shairport-sync` then added shairport-sync to start within rc.local (as a quick fix). Rebooted the pi, and everything worked as expected. I am still trying to figure out why this is the case but I believe the issue to be fixed and have an easy enough work around that it didn't need to continue to be an active issue.

**Comment by rkam** [comment]
Ah, oops!  Sorry, I should have read closer/deeper - I saw the script and the later comment about stopping shairport-sync and just put the two together.

**Comment by bar1792** [comment]
@rkam No worries ;). 

As a side note, I was able to fix the issue with being able to run the script when shairport-sync is enabled by systemctl.

I am unsure if this is ok to do, but I added:

`shairport-sync ALL=(ALL) NOPASSWD:ALL`

to sudoers, by means of `sudo visudo`

So it was as @mikebrady had said earlier, that sudo was looking for a password.

EDIT:
Additionally,(and what I chose to go with instead of `sudo visudo`) the kill process can be accomplished as follows:

`echo "password" | sudo -S pkill arecord`

**Comment by bar1792** [comment]
In the name of being complete with this issue, I have found a complete fix that requires no `echo`ing of passwords nor does it require any change of permissions.

It is important to note that when `systemd` is in charge of starting `shairport-sync`  (by means of `sudo systemctl enable shairport-sync`) `systemd` runs very early on in the boot process, in addition to the early runtime, `systemd` requires a stable and consistent environment to run in. As such when `systemd` runs `shairport-sync` it is run as root in a "non-login shell" aka "non-interactive shell", so there is no TTY. In addition to this, `/etc/sudoers` has not been loaded and therefore `sudo` rules have not been set. Each time `sudo` is used inside a script that is run from `shairport-sync` it will require a password. However, since `systemd` in turn runs `shairport-sync` and `systemd` runs it as root, every child process under the parent process will inherit the same permissions and group settings. Meaning that any scripts that `shairport-sync` runs will run as root and will not require the use of sudo.

With all that said and done, all that needs to be done for this script to be run correctly is change the first line after the shebang from
`/bin/ps -ef | /bin/grep are | /bin/grep -v /bin/grep | /usr/bin/awk '{print $2}'| /usr/bin/xargs /usr/bin/sudo /bin/kill
`
to the following line of code:
`/bin/pkill arecord`


TLDR;
Moral of the story is be stingy with your use of `sudo` if it doesn't work without `sudo` then try `sudo`.

Another thing I learned is that just because `systemd` and `/etc/rc.local` run `shairport-sync` as root does not mean that `shairport-sync` will run in the same exact way.


@mikebrady , I just wanted to include you here just incase this problem happens to someone else. There is a fix, and its quite simple.
